diff --git a/src/portable/synopsys/dwc2/dwc2_esp32.h b/src/portable/synopsys/dwc2/dwc2_esp32.h
index 49b8c54cb..8b5e99741 100644
--- a/src/portable/synopsys/dwc2/dwc2_esp32.h
+++ b/src/portable/synopsys/dwc2/dwc2_esp32.h
@@ -70,11 +70,13 @@ static void dwc2_int_handler_wrap(void* arg) {
   const tusb_role_t role = (tusb_role_t) tu_u16_high((uint16_t)(uintptr_t)arg);
 #if CFG_TUD_ENABLED
   if (role == TUSB_ROLE_DEVICE) {
+    void dcd_int_handler(uint8_t rhport);
     dcd_int_handler(rhport);
   }
 #endif
 #if CFG_TUH_ENABLED && !CFG_TUH_MAX3421
   if (role == TUSB_ROLE_HOST) {
+    void hcd_int_handler(uint8_t rhport, bool in_isr);
     hcd_int_handler(rhport, true);
   }
 #endif
